docs/src/p4est_interfacing.md

```diff
@@ -0,0 +1,333 @@
+```C
+/** Callback function prototype to decide for refinement.
+ * \param [in] p4est       the forest
+ * \param [in] which_tree  the tree containing \a quadrant
+ * \param [in] quadrant    the quadrant that may be refined
+ * \return nonzero if the quadrant shall be refined.
+ */
+typedef int (*p4est_refine_t) (p4est_t * p4est,
+                               p4est_topidx_t which_tree,
+                               p4est_quadrant_t * quadrant);
+
+/** Callback function prototype to decide for refinement.
+ * \param [in] p8est       the forest
+ * \param [in] which_tree  the tree containing \a quadrant
+ * \param [in] quadrant    the quadrant that may be refined
+ * \return nonzero if the quadrant shall be refined.
+ */
+typedef int (*p8est_refine_t) (p8est_t * p8est,
+                               p4est_topidx_t which_tree,
+                               p8est_quadrant_t * quadrant);
+```
+which is implemented as 
+```julia
+function refine_fn(p4est, which_tree, quadrant)
+    # Controller value has been copied to the quadrant's user data storage before.
+    # Unpack quadrant's user data ([global quad ID, controller_value]).
+    # Use `unsafe_load` here since `quadrant.p.user_data isa Ptr{Ptr{Nothing}}`
+    # and we only need the first (only!) entry
+    pw = PointerWrapper(Int, unsafe_load(quadrant.p.user_data))
+    controller_value = pw[2]
+
+    if controller_value > 0
+        # return true (refine)
+        return Cint(1)
+    else
+        # return false (don't refine)
+        return Cint(0)
+    end
+end
+```
+Again, dimensional dispatching is used to handle both 2D and 3D meshes:
+```julia
+# 2D
+function cfunction(::typeof(refine_fn), ::Val{2})
+    @cfunction(refine_fn, 
+               # Note the matching signature to the C function
+               Cint, # Return type
+               # Parameters
+               (Ptr{p4est_t}, Ptr{p4est_topidx_t}, Ptr{p4est_quadrant_t}))
+end
+# 3D
+function cfunction(::typeof(refine_fn), ::Val{3})
+    @cfunction(refine_fn, 
+               # Note the matching signature to the C function
+               Cint, # Return type
+               # Parameters
+               (Ptr{p8est_t}, Ptr{p4est_topidx_t}, Ptr{p8est_quadrant_t}))
+end
+```
+These are then handed over to `refine_p4est!` which eventually calls `refine_fn`.
+
+## Custom Mesh Partitioning
+
+Analogous to the AMR example above, we now discuss how to implement a custom mesh partitioning based on a weighting function/array.
+This might be required for distributed memory (i.e., `MPI`-parallelized simulations) to achieve a more uniform distribution of the computational work across the different processes.
+This is important to maintain scalability and to avoid load imbalance.
```

Comment:
```suggestion
This might be required for distributed memory (i.e., MPI-parallelized simulations) to achieve a more uniform distribution of the computational work across the different processes.
This is important to maintain scalability and to avoid load imbalance.
```